utils/background_disposer.cc

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2025-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+ */
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/on_internal_error.hh>
+#include <seastar/core/shard_id.hh>
+
+#include "utils/background_disposer.hh"
+#include "utils/assert.hh"
+#include "utils/log.hh"
+
+namespace utils {
+
+static logging::logger logger("background_disposer");
+
+background_disposer::basic_item::basic_item(basic_item&& o) {
+    if (_shard != o._shard) {
+        on_fatal_internal_error(logger, format("background_disposer::basic_item[{}]: cannot move from basic_item shard={} to this shard", _shard, o._shard));
+    }
+}
+
+background_disposer::basic_item::~basic_item() {
+    if (this_shard_id() != _shard) {
+        on_fatal_internal_error(logger, format("background_disposer::basic_item[{}]: detroyed on wrong shard", _shard));
+    }
+}
+
+background_disposer::background_disposer() {
+    _done = disposer();
+}
+
+background_disposer::background_disposer(background_disposer&& o) {
+    if (_shard != o._shard) {
+        on_fatal_internal_error(logger, format("background_disposer[{}]: cannot move from disposer shard={} to this shard", _shard, o._shard));
+    }
+    _done = std::exchange(o._done, make_ready_future());
+}
+
+background_disposer::~background_disposer() {
+    SCYLLA_ASSERT(_stopped);
+}
+
+future<> background_disposer::stop() noexcept {
+    if (this_shard_id() != _shard) {
+        on_fatal_internal_error(logger, format("background_disposer[{}]: cannot stop disposer on this shard", _shard));
+    }
+    logger.debug("disposer: stopping");
+    _stopped = true;
+    _cond.signal();
+    co_await std::exchange(_done, make_ready_future());
+    logger.debug("disposer: stopped");
+}
+
+void background_disposer::dispose(std::unique_ptr<basic_item> item) noexcept {
+    if (this_shard_id() != _shard) {
+        on_fatal_internal_error(logger, format("background_disposer[{}]: cannot dispose item={} on this shard", _shard, fmt::ptr(item.get())));
+    }
+    if (this_shard_id() != item->_shard) {
+        on_fatal_internal_error(logger, format("background_disposer[{}]: cannot dispose item={} that was allocated on a foreign shard={}", _shard, fmt::ptr(item.get()), item->_shard));
+    }
+    logger.debug("dispose: queuing item={} for disposal", fmt::ptr(item.get()));
+    try {
+        _items.push_back(std::move(item));
```

Comment:
Alternative 1: the item is always a member of a list, either a live list or a to-be-disposed list, and you use list::splice to move from one list to another, no allocations needed.
Alternative 2: intrusive list.

Reply:
Intrusive list was my first try, but as you said, we need to keep the object alive so the disposer becomes a factory for the objects with life-cycle management.
It's possible and could be used both for unique ptrs and shard ptrs, but it's more complicated.

Reply:
Then alternative 1 should work.

But maybe synchronous destruction on exception isn't too bad, it's unlikely to ever happen.

Reply:
used intrusive_list in v4